### PR TITLE
State changes read operations

### PR DESCRIPTION
**Pull request by ssd04**
## Reasoning behind the pull request
- Added config field for enabling read state changes
- Separate components for write state changes and data analysis state changes

## Testing procedure
- Standard system test with outport driver enabled

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: state/stateChanges/writeCollector.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+	logger "github.com/multiversx/mx-chain-logger-go"
+	vmcommon "github.com/multiversx/mx-chain-vm-common-go"
+)
+
+var log = logger.GetOrCreate("state/stateChanges")
+
+// ErrStateChangesIndexOutOfBounds signals that the state changes index is out of bounds
+var ErrStateChangesIndexOutOfBounds = errors.New("state changes index out of bounds")
+
+type StateChange interface {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/stateChanges/writeCollector.go:21]: And move to the interface file

File: state/stateChanges/dataAnalysisCollector.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+// NewDataAnalysisStateChangesCollector will create a new instance of data analysis collector
+func NewDataAnalysisStateChangesCollector(storer storage.Persister) (*dataAnalysisCollector, error) {
+	if check.IfNil(storer) {
+		return nil, storage.ErrNilPersisterFactory
+	}
+
+	return &dataAnalysisCollector{
+		stateChangesCollector: &stateChangesCollector{
```
Comment (BeniaminDrasovean) [inline state/stateChanges/dataAnalysisCollector.go:60]: Use `NewStateChangesCollector()`

File: state/stateChanges/writeCollector.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	defer scc.stateChangesMut.Unlock()
+
+	stateChangesForTxs := make([]StateChangesForTx, 0)
+
+	for i := 0; i < len(scc.stateChanges); i++ {
+		txHash := scc.stateChanges[i].GetTxHash()
+
+		if len(txHash) == 0 {
+			log.Warn("empty tx hash, state change event not associated to a transaction")
+			break
```
[the remark this answers is not in the document]
Comment (ssd04) [reply state/stateChanges/writeCollector.go:79]: changed to use continue

File: epochStart/metachain/systemSCs_test.go [hunk 4]
```diff
@@ -779,9 +780,10 @@ func createFullArgumentsForSystemSCProcessing(enableEpochsConfig config.EnableEp
 
 	trieFactoryManager, _ := trie.CreateTrieStorageManager(storageManagerArgs, storageMock.GetStorageManagerOptions())
 	argsAccCreator := factory.ArgsAccountCreator{
-		Hasher:              hasher,
-		Marshaller:          marshalizer,
-		EnableEpochsHandler: &enableEpochsHandlerMock.EnableEpochsHandlerStub{},
+		Hasher:                hasher,
+		Marshaller:            marshalizer,
+		EnableEpochsHandler:   &enableEpochsHandlerMock.EnableEpochsHandlerStub{},
+		StateChangesCollector: stateChanges.NewStateChangesCollector(),
```
[the remark this answers is not in the document]
Comment (ssd04) [reply epochStart/metachain/systemSCs_test.go:786]: done

File: factory/state/stateComponents.go [hunk 5]
```diff
@@ -236,9 +273,9 @@ func (scf *stateComponentsFactory) createPeerAdapter(triesContainer common.Tries
 	}
 
 	stateChangesCollector := disabled.NewDisabledStateChangesCollector()
-	if scf.config.StateTriesConfig.CollectStateChangesEnabled {
-		stateChangesCollector = state.NewStateChangesCollector()
-	}
+	// if scf.config.StateTriesConfig.CollectStateChangesEnabled {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply factory/state/stateComponents.go:276]: updated

File: outport/host/errors.go [hunk 6]
```diff
@@ -7,3 +7,6 @@ var ErrHostIsClosed = errors.New("server is closed")
 
 // ErrNilHost signals that a nil host has been provided
 var ErrNilHost = errors.New("nil host provided")
+
+// ErrNilStateChangesCollector signals that a nil state change collector has been provided.
+var ErrNilStateChangesCollector = errors.New("nil state changes collector provided")
```
[the remark this answers is not in the document]
Comment (ssd04) [reply outport/host/errors.go:12]: removed

File: state/accountsDB.go [hunk 7]
```diff
@@ -273,17 +285,20 @@ func (adb *AccountsDB) SaveAccount(account vmcommon.AccountHandler) error {
...
+	stateChange := &stateChange.StateChange{
+		Type:            "write",
 		MainTrieKey:     account.AddressBytes(),
 		MainTrieVal:     marshalledAccount,
 		DataTrieChanges: newDataTrieValues,
+		Operation:       "saveAccount",
 	}
-	adb.stateChangesCollector.AddStateChange(stateChange)
+
+	adb.stateChangesCollector.AddSaveAccountStateChange(oldAccount, account, stateChange)
```
[the remark this answers is not in the document]
Comment (ssd04) [reply state/accountsDB.go:296]: done

File: state/accountsDB_test.go [hunk 8]
```diff
@@ -423,127 +426,127 @@ func TestAccountsDB_SaveAccountMalfunctionMarshallerShouldErr(t *testing.T) {
...
-		Key:     []byte("key1"),
-		Address: address,
-	}
-	serializedTrieVal, _ := marshaller.Marshal(trieVal)
-	assert.Equal(t, hasher.Compute("key1"), accountStateChange.DataTrieChanges[0].Key)
-	assert.Equal(t, serializedTrieVal, accountStateChange.DataTrieChanges[0].Val)
-	assert.Equal(t, []byte("key1"), accountStateChange.DataTrieChanges[1].Key)
-	assert.Equal(t, []byte(nil), accountStateChange.DataTrieChanges[1].Val)
-}
+// func TestAccountsDB_SaveAccountCollectsAllStateChanges(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply state/accountsDB_test.go:429]: updated tests

File: state/stateChanges/dataAnalysisCollector.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-core-go/data/transaction"
+	vmcommon "github.com/multiversx/mx-chain-vm-common-go"
+
+	"github.com/multiversx/mx-chain-go/storage"
+)
+
+// TODO: use proto stucts
+type DataAnalysisStateChangeDTO struct {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply state/stateChanges/dataAnalysisCollector.go:17]: changed to be unexported

File: state/stateChanges/dataAnalysisCollector.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+	Balance         bool   `json:"balanceChanged"`
+	CodeHash        bool   `json:"codeHashChanged"`
+	RootHash        bool   `json:"rootHashChanged"`
+	DeveloperReward bool   `json:"developerRewardChanged"`
+	OwnerAddress    bool   `json:"ownerAddressChanged"`
+	UserName        bool   `json:"userNameChanged"`
+	CodeMetadata    bool   `json:"codeMetadataChanged"`
+}
+
+type DataAnalysisStateChangesForTx struct {
```
[the remark this answers is not in the document]
Comment (ssd04) [reply state/stateChanges/dataAnalysisCollector.go:30]: changed to be unexported

File: state/stateChanges/dataAnalysisCollector_test.go [hunk 11]
```diff
@@ -0,0 +1 @@
+package stateChanges
```
[the remark this answers is not in the document]
Comment (ssd04) [reply state/stateChanges/dataAnalysisCollector_test.go:1]: added todo comment

File: state/stateChanges/writeCollector.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ssd04) [reply state/stateChanges/writeCollector.go:21]: done

File: state/stateChanges/writeCollector.go [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+	data "github.com/multiversx/mx-chain-core-go/data/stateChange"
+	"github.com/multiversx/mx-chain-core-go/data/transaction"
+	logger "github.com/multiversx/mx-chain-logger-go"
+	vmcommon "github.com/multiversx/mx-chain-vm-common-go"
+)
+
+var log = logger.GetOrCreate("state/stateChanges")
+
+// ErrStateChangesIndexOutOfBounds signals that the state changes index is out of bounds
+var ErrStateChangesIndexOutOfBounds = errors.New("state changes index out of bounds")
```
[the remark this answers is not in the document]
Comment (ssd04) [reply state/stateChanges/writeCollector.go:19]: done

File: state/stateChanges/dataAnalysisCollector.go (hunk 2 above)
Comment (ssd04) [reply state/stateChanges/dataAnalysisCollector.go:60]: done

File: state/stateChanges/dataAnalysisCollector.go [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	dataAnalysisStateChangesForTxs := make([]DataAnalysisStateChangesForTx, len(stateChangesForTxs))
+
+	for _, stateChangeForTx := range stateChangesForTxs {
+		txHash := string(stateChangeForTx.TxHash)
+
+		cachedTx, txOk := scc.cachedTxs[string(txHash)]
```
[the remark this answers is not in the document]
Comment (ssd04) [reply state/stateChanges/dataAnalysisCollector.go:144]: done

File: genesis/process/genesisBlockCreator_test.go [hunk 14]
```diff
@@ -200,9 +201,10 @@ func createMockArgument(
 	}
 
 	argsAccCreator := factoryState.ArgsAccountCreator{
-		Hasher:              &hashingMocks.HasherMock{},
-		Marshaller:          &mock.MarshalizerMock{},
-		EnableEpochsHandler: &enableEpochsHandlerMock.EnableEpochsHandlerStub{},
+		Hasher:                &hashingMocks.HasherMock{},
+		Marshaller:            &mock.MarshalizerMock{},
+		EnableEpochsHandler:   &enableEpochsHandlerMock.EnableEpochsHandlerStub{},
+		StateChangesCollector: stateChanges.NewStateChangesCollector(),
```
Comment (cristure) [inline genesis/process/genesisBlockCreator_test.go:207]: any reason why the mock isn't used instead?

File: integrationTests/state/stateTrie/stateTrie_test.go [hunk 15]
```diff
@@ -2732,9 +2734,10 @@ func createAccountsDBTestSetup() *state.AccountsDB {
 	tr, _ := trie.NewTrie(trieStorage, integrationTests.TestMarshalizer, integrationTests.TestHasher, &enableEpochsHandlerMock.EnableEpochsHandlerStub{}, maxTrieLevelInMemory)
 	spm, _ := storagePruningManager.NewStoragePruningManager(ewl, 10)
 	argsAccCreator := factory.ArgsAccountCreator{
-		Hasher:              integrationTests.TestHasher,
-		Marshaller:          integrationTests.TestMarshalizer,
-		EnableEpochsHandler: &enableEpochsHandlerMock.EnableEpochsHandlerStub{},
+		Hasher:                integrationTests.TestHasher,
+		Marshaller:            integrationTests.TestMarshalizer,
+		EnableEpochsHandler:   &enableEpochsHandlerMock.EnableEpochsHandlerStub{},
+		StateChangesCollector: stateChanges.NewStateChangesCollector(),
```
Comment (cristure) [inline integrationTests/state/stateTrie/stateTrie_test.go:2740]: same as above

File: integrationTests/testProcessorNode.go [hunk 16]
```diff
@@ -2234,6 +2235,7 @@ func (tpn *TestProcessorNode) initBlockProcessor() {
 		BlockProcessingCutoffHandler: &testscommon.BlockProcessingCutoffStub{},
 		ManagedPeersHolder:           &testscommon.ManagedPeersHolderStub{},
 		SentSignaturesTracker:        &testscommon.SentSignatureTrackerStub{},
+		StateChangesCollector:        stateDisabled.NewDisabledStateChangesCollector(),
```
Comment (cristure) [inline integrationTests/testProcessorNode.go:2238]: could use the mock instead.

File: outport/process/outportDataProvider_test.go [hunk 17]
```diff
@@ -45,6 +47,7 @@ func createArgOutportDataProvider() ArgOutportDataProvider {
 		ExecutionOrderHandler:    &commonMocks.TxExecutionOrderHandlerStub{},
 		Marshaller:               &marshallerMock.MarshalizerMock{},
 		Hasher:                   &hashingMocks.HasherMock{},
+		StateChangesCollector:    stateChanges.NewStateChangesCollector(),
```
Comment (cristure) [inline outport/process/outportDataProvider_test.go:50]: same here

File: process/rewardTransaction/process_test.go [hunk 18]
```diff
@@ -260,7 +261,13 @@ func TestRewardTxProcessor_ProcessRewardTransactionToASmartContractShouldWork(t
 
 	address := []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6}
 
-	dtt, _ := trackableDataTrie.NewTrackableDataTrie(address, &hashingMocks.HasherMock{}, &marshallerMock.MarshalizerMock{}, enableEpochsHandlerMock.NewEnableEpochsHandlerStub())
+	dtt, _ := trackableDataTrie.NewTrackableDataTrie(
+		address,
+		&hashingMocks.HasherMock{},
+		&marshallerMock.MarshalizerMock{},
+		enableEpochsHandlerMock.NewEnableEpochsHandlerStub(),
+		disabled.NewDisabledStateChangesCollector(),
```
Comment (cristure) [inline process/rewardTransaction/process_test.go:269]: same here

File: process/scToProtocol/stakingToPeer_test.go [hunk 19]
```diff
@@ -61,7 +62,13 @@ func createBlockBody() *block.Body {
 }
 
 func createStakingScAccount() state.UserAccountHandler {
-	dtt, _ := trackableDataTrie.NewTrackableDataTrie(vm.StakingSCAddress, &hashingMocks.HasherMock{}, &marshallerMock.MarshalizerMock{}, enableEpochsHandlerMock.NewEnableEpochsHandlerStub())
+	dtt, _ := trackableDataTrie.NewTrackableDataTrie(
+		vm.StakingSCAddress,
+		&hashingMocks.HasherMock{},
+		&marshallerMock.MarshalizerMock{},
+		enableEpochsHandlerMock.NewEnableEpochsHandlerStub(),
+		stateChanges.NewStateChangesCollector(),
```
Comment (cristure) [inline process/scToProtocol/stakingToPeer_test.go:70]: here as well.

File: process/smartContract/processorV2/process_test.go [hunk 20]
```diff
@@ -63,9 +64,10 @@ func createMockPubkeyConverter() *testscommon.PubkeyConverterMock {
 
 func createAccount(address []byte) state.UserAccountHandler {
 	argsAccCreation := stateFactory.ArgsAccountCreator{
-		Hasher:              &hashingMocks.HasherMock{},
-		Marshaller:          &marshallerMock.MarshalizerMock{},
-		EnableEpochsHandler: enableEpochsHandlerMock.NewEnableEpochsHandlerStub(),
+		Hasher:                &hashingMocks.HasherMock{},
+		Marshaller:            &marshallerMock.MarshalizerMock{},
+		EnableEpochsHandler:   enableEpochsHandlerMock.NewEnableEpochsHandlerStub(),
+		StateChangesCollector: stateChanges.NewStateChangesCollector(),
```
Comment (cristure) [inline process/smartContract/processorV2/process_test.go:70]: same here

File: state/accountsDB_test.go [hunk 21]
```diff
@@ -149,9 +154,10 @@ func getDefaultStateComponents(
 	ewl, _ := evictionWaitingList.NewMemoryEvictionWaitingList(ewlArgs)
 	spm, _ := storagePruningManager.NewStoragePruningManager(ewl, generalCfg.PruningBufferLen)
 	argsAccCreator := factory.ArgsAccountCreator{
-		Hasher:              hasher,
-		Marshaller:          marshaller,
-		EnableEpochsHandler: enableEpochsHandler,
+		Hasher:                hasher,
+		Marshaller:            marshaller,
+		EnableEpochsHandler:   enableEpochsHandler,
+		StateChangesCollector: stateDisabled.NewDisabledStateChangesCollector(),
```
Comment (cristure) [inline state/accountsDB_test.go:160]: here as well.

File: state/accountsDB_test.go [hunk 22]
```diff
@@ -175,7 +181,7 @@ func getDefaultStateComponents(
 		StoragePruningManager: spm,
 		AddressConverter:      &testscommon.PubkeyConverterMock{},
 		SnapshotsManager:      snapshotsManager,
-		StateChangesCollector: state.NewStateChangesCollector(),
+		StateChangesCollector: stateChanges.NewStateChangesCollector(),
```
Comment (cristure) [inline state/accountsDB_test.go:184]: here as well.

File: state/accountsDB_test.go [hunk 23]
```diff
@@ -1919,9 +1932,10 @@ func TestAccountsDB_MainTrieAutomaticallyMarksCodeUpdatesForEviction(t *testing.
 	argsAccountsDB.Hasher = hasher
 	argsAccountsDB.Marshaller = marshaller
 	argsAccCreator := factory.ArgsAccountCreator{
-		Hasher:              hasher,
-		Marshaller:          marshaller,
-		EnableEpochsHandler: &enableEpochsHandlerMock.EnableEpochsHandlerStub{},
+		Hasher:                hasher,
+		Marshaller:            marshaller,
+		EnableEpochsHandler:   &enableEpochsHandlerMock.EnableEpochsHandlerStub{},
+		StateChangesCollector: stateChanges.NewStateChangesCollector(),
```
Comment (cristure) [inline state/accountsDB_test.go:1938]: here as well.

File: state/accountsDB_test.go [hunk 24]
```diff
@@ -2000,9 +2014,10 @@ func TestAccountsDB_RemoveAccountMarksObsoleteHashesForEviction(t *testing.T) {
...
```
Comment (cristure) [inline state/accountsDB_test.go:2020]: here as well.

File: state/accountsDB_test.go [hunk 25]
```diff
@@ -2202,9 +2217,10 @@ func TestAccountsDB_GetCode(t *testing.T) {
...
```
Comment (cristure) [inline state/accountsDB_test.go:2223]: here as well.

File: state/accountsDB_test.go [hunk 26]
```diff
@@ -2369,9 +2385,10 @@ func TestAccountsDB_Close(t *testing.T) {
...
```
Comment (cristure) [inline state/accountsDB_test.go:2391]: here as well.

File: state/accountsDB_test.go [hunk 27]
```diff
@@ -2657,9 +2674,10 @@ func BenchmarkAccountsDb_GetCodeEntry(b *testing.B) {
...
```
Comment (cristure) [inline state/accountsDB_test.go:2680]: here as well.

File: state/accountsDB_test.go [hunk 28]
```diff
@@ -2977,9 +2997,10 @@ func TestAccountsDB_RevertTxWhichMigratesDataRemovesMigratedData(t *testing.T) {
 	argsAccountsDB.Hasher = hasher
 	argsAccountsDB.Marshaller = marshaller
 	argsAccCreator := factory.ArgsAccountCreator{
-		Hasher:              hasher,
-		Marshaller:          marshaller,
-		EnableEpochsHandler: enableEpochsHandler,
+		Hasher:                hasher,
+		Marshaller:            marshaller,
+		EnableEpochsHandler:   enableEpochsHandler,
+		StateChangesCollector: stateChanges.NewStateChangesCollector(),
```
Comment (cristure) [inline state/accountsDB_test.go:3003]: here as well.

File: genesis/process/genesisBlockCreator_test.go (hunk 14 above)
Comment (ssd04) [reply genesis/process/genesisBlockCreator_test.go:207]: changed to use mock

File: state/accountsDB_test.go (hunk 22 above)
Comment (ssd04) [reply state/accountsDB_test.go:184]: here is intentional

File: state/accountsDB_test.go (hunk 28 above)
Comment (ssd04) [reply state/accountsDB_test.go:3003]: updated to use mock when suitable

**Review by cristure — APPROVED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

**Pull request merged**